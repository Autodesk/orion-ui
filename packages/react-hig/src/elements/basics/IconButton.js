/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as HIG from 'hig.web';

import HIGElement from '../HIGElement';
import * as PropTypes from 'prop-types';
import createComponent from '../../adapters/createComponent';

class IconButton extends HIGElement {
  constructor(initialProps) {
    super(HIG.IconButton, initialProps);
  }

  componentDidMount() {
    this.commitUpdate(this.props);
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    this.processUpdateProps(updatePayload)
      .mapToHIGFunctions({
        title: 'setTitle',
        link: 'setLink',
        icon: 'setIcon'
      })
      .mapToHIGEventListeners(['onClick', 'onHover', 'onFocus', 'onBlur'])
      .handle('disabled', value => {
        value ? this.hig.disable() : this.hig.enable();
      });
  }
}

const IconButtonComponent = createComponent(IconButton);

IconButton.propTypes = {
  title: PropTypes.string,
  icon: PropTypes.string,
  disabled: PropTypes.bool,
  link: PropTypes.string,
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onHover: PropTypes.func
};

IconButtonComponent.__docgenInfo = {
  props: {
    title: {
      description: 'sets the title of a button'
    },

    link: {
      description: 'sets the link of a button'
    },

    icon: {
      description: 'specifies icon for button'
    },

    onClick: {
      description: 'triggers when you click the button'
    },

    onHover: {
      description: 'triggers when you hover over the button'
    },

    onFocus: {
      description: 'triggers focus is moved to button'
    },

    onBlur: {
      description: 'triggers blur when focuss is moved away from icon'
    }
  }
};

export default IconButtonComponent;
