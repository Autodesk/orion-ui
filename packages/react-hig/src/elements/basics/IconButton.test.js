/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import IconButton from './IconButton';

describe('<IconButton>', () => {
  /**
     * Creates a hig.web button and returns the instance and it's container
     *
     * @param {object} defaults hig.web defaults for the button
     */
  function createHigButton(defaults = {}) {
    const higContainer = document.createElement('div');

    // use spread here to clone defaults since HIG.Button mutates this object
    const higButton = new HIG.IconButton({ ...defaults });

    higButton.mount(higContainer);

    return { higButton, higContainer };
	}
	
	 it('renders the icon button', () => {
    const defaults = { title: 'icon button', link: 'http://example.com', icon: 'gear',  };

		const { higButton, higContainer } = createHigButton(defaults);
		const container = document.createElement('div');
		higButton.setIcon('gear');

    const wrapper = mount(<IconButton {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the disabled icon button', () => {
    const defaults = { title: 'icon button', link: 'http://example.com', icon: 'gear' };

    const { higButton, higContainer } = createHigButton(defaults);
		const container = document.createElement('div');
		higButton.setIcon('gear');
		higButton.disable();

    const wrapper = mount(<IconButton {...defaults} disabled={true} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });
});
