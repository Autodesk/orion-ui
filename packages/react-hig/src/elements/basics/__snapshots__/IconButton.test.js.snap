// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton> renders the disabled icon button 1`] = `
"<a href=\\"http://example.com\\" tabindex=\\"-1\\" class=\\"hig__icon-button hig__icon-button--disabled\\" title=\\"icon button\\">
    <div class=\\"hig__icon-button__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"gear\\"><path d=\\"M19.9368983,10.4543901 C19.5951319,10.4052173 19.2316926,10.0976789 19.1258284,9.77097137 L18.6181803,8.53665041 C18.4589672,8.23244572 18.4964781,7.75738634 18.7015379,7.4831854 L19.5267786,6.38471476 C19.7318385,6.10968038 19.7160005,5.67295912 19.4876006,5.41626037 L18.5840037,4.51114723 C18.3264285,4.28445222 17.8904681,4.26695004 17.6137207,4.47364254 L16.516734,5.2970788 C16.2416538,5.5037713 15.7681824,5.54127599 15.4630934,5.38125599 L14.2294002,4.87369254 C13.9001375,4.76867942 13.5925478,4.40530066 13.5458675,4.06442472 L13.3508107,2.70675501 C13.3032968,2.36587907 12.9832034,2.05834063 12.6422707,2.02166938 C12.6422707,2.02166938 12.4297087,2 12.0004168,2 C11.5719585,2 11.3593965,2.02166938 11.3593965,2.02166938 C11.0176301,2.05834063 10.6975368,2.36587907 10.6491893,2.70675501 L10.454966,4.06442472 C10.4066186,4.40530066 10.0990289,4.76867942 9.77143333,4.87369254 L8.53774017,5.38125599 C8.23265119,5.54044256 7.75917976,5.5037713 7.48409953,5.2970788 L6.38544576,4.4728091 C6.11036552,4.2661166 5.67357146,4.28278535 5.41599633,4.51031379 L4.51239945,5.41542693 C4.28566665,5.67295912 4.26732797,6.10884694 4.47405493,6.38388132 L5.29762847,7.48235196 C5.50518901,7.75738634 5.54186638,8.23161228 5.38265327,8.53581698 L4.87333806,9.77097137 C4.769141,10.0985123 4.40486809,10.4060508 4.06393531,10.4543901 L2.70603926,10.6469142 C2.36594007,10.696087 2.05835035,11.0152936 2.02250656,11.3578364 C2.02250656,11.3578364 2,11.570363 2,11.9995833 C2,12.4288036 2.02250656,12.6413302 2.02250656,12.6413302 C2.05835035,12.983873 2.36594007,13.3022461 2.70603926,13.3514189 L4.06393531,13.5447764 C4.40403451,13.5939492 4.769141,13.9014877 4.87333806,14.2281952 L5.38265327,15.4616827 C5.54186638,15.7658874 5.50435544,16.2417802 5.29762847,16.5151477 L4.47405493,17.6136184 C4.26732797,17.8886527 4.21731338,18.2628662 4.36402284,18.4420553 C4.50989872,18.6229112 4.9283541,19.0688003 4.92918768,19.0704671 C4.93002126,19.072134 5.07172925,19.2004834 5.24177885,19.3571696 C5.41182845,19.5138559 6.11036552,19.7313831 6.38544576,19.5263575 L7.48326595,18.7012543 C7.75834618,18.4962287 8.23181761,18.4570571 8.5369066,18.6179106 L9.77059976,19.125474 C10.0990289,19.2296537 10.4066186,19.5938659 10.4541325,19.9347418 L10.6483558,21.2924116 C10.6967032,21.6324541 11.015963,21.9416594 11.3585629,21.9774972 C11.3585629,21.9774972 11.5711249,22 11.9995832,22 C12.4288751,22 12.6414371,21.9774972 12.6414371,21.9774972 C12.9823699,21.9416594 13.3024632,21.6324541 13.3499771,21.2924116 L13.545034,19.9347418 C13.5925478,19.5946993 13.9001375,19.2296537 14.2285667,19.125474 L15.4622598,18.6162437 C15.7665152,18.4570571 16.2408202,18.4945618 16.5159005,18.7012543 L17.6145542,19.5263575 C17.8913016,19.7313831 18.3264285,19.7155478 18.5848372,19.4871859 L19.486767,18.5837396 C19.7143333,18.3245406 19.7310049,17.8886527 19.5259451,17.6136184 L18.7007044,16.5151477 C18.4956446,16.2417802 18.4581336,15.7667208 18.6173467,15.4616827 L19.1249948,14.2281952 C19.230859,13.9006542 19.5934648,13.5931158 19.9360647,13.5447764 L21.2922936,13.3514189 C21.6340599,13.3022461 21.9416496,12.983873 21.9774934,12.6413302 C21.9774934,12.6413302 22,12.4288036 22,11.9995833 C22,11.570363 21.9774934,11.3578364 21.9774934,11.3578364 C21.9416496,11.016127 21.6340599,10.696087 21.2922936,10.6469142 L19.9368983,10.4543901 Z M12.0004168,15.4121833 C10.1152098,15.4121833 8.58724779,13.8833385 8.58724779,11.9995833 C8.58724779,10.1146905 10.1152098,8.58698322 12.0004168,8.58698322 C13.8856238,8.58698322 15.4135858,10.115828 15.4135858,11.9995833 C15.4135858,13.8833385 13.8844861,15.4121833 12.0004168,15.4121833 Z\\" transform=\\"translate(-2 -2)\\" fill-rule=\\"nonzero\\" fill=\\"#353636\\"></path></svg></div></div>
</a>"
`;

exports[`<IconButton> renders the icon button with initial props 1`] = `
"<a href=\\"http://example.com\\" tabindex=\\"0\\" class=\\"hig__icon-button\\" title=\\"icon button\\">
    <div class=\\"hig__icon-button__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"gear\\"><path d=\\"M19.9368983,10.4543901 C19.5951319,10.4052173 19.2316926,10.0976789 19.1258284,9.77097137 L18.6181803,8.53665041 C18.4589672,8.23244572 18.4964781,7.75738634 18.7015379,7.4831854 L19.5267786,6.38471476 C19.7318385,6.10968038 19.7160005,5.67295912 19.4876006,5.41626037 L18.5840037,4.51114723 C18.3264285,4.28445222 17.8904681,4.26695004 17.6137207,4.47364254 L16.516734,5.2970788 C16.2416538,5.5037713 15.7681824,5.54127599 15.4630934,5.38125599 L14.2294002,4.87369254 C13.9001375,4.76867942 13.5925478,4.40530066 13.5458675,4.06442472 L13.3508107,2.70675501 C13.3032968,2.36587907 12.9832034,2.05834063 12.6422707,2.02166938 C12.6422707,2.02166938 12.4297087,2 12.0004168,2 C11.5719585,2 11.3593965,2.02166938 11.3593965,2.02166938 C11.0176301,2.05834063 10.6975368,2.36587907 10.6491893,2.70675501 L10.454966,4.06442472 C10.4066186,4.40530066 10.0990289,4.76867942 9.77143333,4.87369254 L8.53774017,5.38125599 C8.23265119,5.54044256 7.75917976,5.5037713 7.48409953,5.2970788 L6.38544576,4.4728091 C6.11036552,4.2661166 5.67357146,4.28278535 5.41599633,4.51031379 L4.51239945,5.41542693 C4.28566665,5.67295912 4.26732797,6.10884694 4.47405493,6.38388132 L5.29762847,7.48235196 C5.50518901,7.75738634 5.54186638,8.23161228 5.38265327,8.53581698 L4.87333806,9.77097137 C4.769141,10.0985123 4.40486809,10.4060508 4.06393531,10.4543901 L2.70603926,10.6469142 C2.36594007,10.696087 2.05835035,11.0152936 2.02250656,11.3578364 C2.02250656,11.3578364 2,11.570363 2,11.9995833 C2,12.4288036 2.02250656,12.6413302 2.02250656,12.6413302 C2.05835035,12.983873 2.36594007,13.3022461 2.70603926,13.3514189 L4.06393531,13.5447764 C4.40403451,13.5939492 4.769141,13.9014877 4.87333806,14.2281952 L5.38265327,15.4616827 C5.54186638,15.7658874 5.50435544,16.2417802 5.29762847,16.5151477 L4.47405493,17.6136184 C4.26732797,17.8886527 4.21731338,18.2628662 4.36402284,18.4420553 C4.50989872,18.6229112 4.9283541,19.0688003 4.92918768,19.0704671 C4.93002126,19.072134 5.07172925,19.2004834 5.24177885,19.3571696 C5.41182845,19.5138559 6.11036552,19.7313831 6.38544576,19.5263575 L7.48326595,18.7012543 C7.75834618,18.4962287 8.23181761,18.4570571 8.5369066,18.6179106 L9.77059976,19.125474 C10.0990289,19.2296537 10.4066186,19.5938659 10.4541325,19.9347418 L10.6483558,21.2924116 C10.6967032,21.6324541 11.015963,21.9416594 11.3585629,21.9774972 C11.3585629,21.9774972 11.5711249,22 11.9995832,22 C12.4288751,22 12.6414371,21.9774972 12.6414371,21.9774972 C12.9823699,21.9416594 13.3024632,21.6324541 13.3499771,21.2924116 L13.545034,19.9347418 C13.5925478,19.5946993 13.9001375,19.2296537 14.2285667,19.125474 L15.4622598,18.6162437 C15.7665152,18.4570571 16.2408202,18.4945618 16.5159005,18.7012543 L17.6145542,19.5263575 C17.8913016,19.7313831 18.3264285,19.7155478 18.5848372,19.4871859 L19.486767,18.5837396 C19.7143333,18.3245406 19.7310049,17.8886527 19.5259451,17.6136184 L18.7007044,16.5151477 C18.4956446,16.2417802 18.4581336,15.7667208 18.6173467,15.4616827 L19.1249948,14.2281952 C19.230859,13.9006542 19.5934648,13.5931158 19.9360647,13.5447764 L21.2922936,13.3514189 C21.6340599,13.3022461 21.9416496,12.983873 21.9774934,12.6413302 C21.9774934,12.6413302 22,12.4288036 22,11.9995833 C22,11.570363 21.9774934,11.3578364 21.9774934,11.3578364 C21.9416496,11.016127 21.6340599,10.696087 21.2922936,10.6469142 L19.9368983,10.4543901 Z M12.0004168,15.4121833 C10.1152098,15.4121833 8.58724779,13.8833385 8.58724779,11.9995833 C8.58724779,10.1146905 10.1152098,8.58698322 12.0004168,8.58698322 C13.8856238,8.58698322 15.4135858,10.115828 15.4135858,11.9995833 C15.4135858,13.8833385 13.8844861,15.4121833 12.0004168,15.4121833 Z\\" transform=\\"translate(-2 -2)\\" fill-rule=\\"nonzero\\" fill=\\"#353636\\"></path></svg></div></div>
</a>"
`;

exports[`<IconButton> renders the icon button with updated props 1`] = `
"<a href=\\"http://example.com\\" tabindex=\\"-1\\" class=\\"hig__icon-button hig__icon-button--disabled\\" title=\\"icon button\\">
    <div class=\\"hig__icon-button__icon\\"><div class=\\"hig__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"assets\\"><title>hig-light/globalnav/sidenav/icon/assets</title><g fill=\\"#221F20\\" fill-rule=\\"evenodd\\"><path d=\\"M10.1898101,0.916614844 L9.67284723,8.01491201 L9.15588441,0.916614844 C9.15588441,0.916614844 7.37499587,-0.81644658 2.12744709,0.48294645 L2.12744709,12.0829042 C2.12744709,12.0829042 6.89296297,11.6014087 8.33744951,12.3725539 C8.33744951,12.3725539 8.29284669,12.9427177 9.66048741,12.950241 L9.67284723,12.950241 L9.68574443,12.950241 C11.0528478,12.9427177 11.0082449,12.3725539 11.0082449,12.3725539 C12.4527315,11.6014087 17.2182474,12.0829042 17.2182474,12.0829042 L17.2182474,0.48294645 C11.9706986,-0.81644658 10.1898101,0.916614844 10.1898101,0.916614844\\" transform=\\"translate(2.327 4.245)\\"></path><path d=\\"M0.91156344,2.55923464 L0.885231654,13.9028606 L0.455862327,13.4729539 L7.29245378,13.4192155 C7.34887904,13.421365 7.39186971,13.4697296 7.38972017,13.52508 C7.38649587,13.5777436 7.3435052,13.6196595 7.29245378,13.6218091 L0.455862327,14.3333047 C0.229623921,14.3440523 0.0372406687,14.1694027 0.0270303844,13.9442391 L0.0259556176,13.925968 L0.0259556176,13.9028606 L0.000161215016,2.55923464 C-0.000376168371,2.3082766 0.203292135,2.10353353 0.453712794,2.10245876 C0.705745602,2.101384 0.911026056,2.30612707 0.91156344,2.55708511 L0.91156344,2.55923464 Z\\" transform=\\"translate(2.327 4.245)\\"></path><path d=\\"M19.3458019,2.55923464 L19.3200075,13.9028606 L19.3200075,13.9248933 C19.3183954,14.1516691 19.1351477,14.3343794 18.9089092,14.3343794 L18.8895634,14.3333047 L12.0535094,13.6218091 C11.9970841,13.6196595 11.9540934,13.571295 11.956243,13.5159445 C11.9594673,13.463281 12.002458,13.421365 12.0535094,13.4192155 L18.8895634,13.4729539 L18.4607315,13.9028606 L18.4343997,2.55923464 C18.4333249,2.3082766 18.6375306,2.10353353 18.8879513,2.10245876 C19.1405215,2.101384 19.3452646,2.30612707 19.3458019,2.55708511 L19.3458019,2.55923464 Z\\" transform=\\"translate(2.327 4.245)\\"></path></g></svg></div></div>
</a>"
`;
